src/components/MagicCodeInput.js

```diff
@@ -0,0 +1,286 @@
+        this.onKeyPress = this.onKeyPress.bind(this);
+    }
+
+    componentDidMount() {
+        if (!this.props.forwardedRef) {
+            return;
+        }
+        this.props.forwardedRef(this.inputRef);
+
+        if (!this.props.autoFocus) {
+            return;
+        }
+
+        if (this.props.shouldDelayFocus) {
+            this.focusTimeout = setTimeout(() => this.inputRef.focus(), CONST.ANIMATED_TRANSITION);
+            return;
+        }
+        this.inputRef.focus();
+    }
+
+    componentDidUpdate(prevProps) {
+        if (prevProps.value === this.props.value) {
+            return;
+        }
+
+        this.setState({
+            numbers: this.decomposeString(this.props.value),
+        });
+    }
+
+    componentWillUnmount() {
+        if (!this.focusTimeout) {
+            return;
+        }
+        clearTimeout(this.focusTimeout);
+    }
+
+    /**
+     * Focuses on the input when it is pressed.
+     *
+     * @param {Object} event
+     * @param {Number} index
+     */
+    onFocus(event, index) {
+        event.preventDefault();
+        this.setState({
+            input: '',
+            focusedIndex: index,
+            editIndex: index,
+        });
+        this.inputRef.focus();
+    }
+
+    /**
+     * Updates the magic inputs with the contents written in the
+     * input. It spreads each number into each input and updates
+     * the focused input on the next empty one, if exists.
+     * It handles both fast typing and only one digit at a time
+     * in a specific position.
+     *
+     * @param {String} value
+     */
+    onChangeText(value) {
+        if (_.isUndefined(value) || _.isEmpty(value) || !ValidationUtils.isNumeric(value)) {
+            return;
+        }
+
+        const numbersArr = value.trim().split('');
+        this.setState((prevState) => {
+            const numbers = [
+                ...prevState.numbers.slice(0, prevState.editIndex),
+                ...numbersArr.slice(0, CONST.MAGIC_CODE_LENGTH - prevState.editIndex),
+            ];
+
+            // Updates the focused input taking into consideration the last input
+            // edited and the number of digits added by the user.
+            const focusedIndex = Math.min(prevState.editIndex + (numbersArr.length - 1), CONST.MAGIC_CODE_LENGTH - 1);
+
+            return {
+                numbers,
+                focusedIndex,
+                input: value,
+                editIndex: prevState.editIndex,
+            };
+        }, () => {
+            const finalInput = this.composeToString(this.state.numbers);
+            this.props.onChange(finalInput);
+
+            // If the input is complete and submit on complete is enabled, waits for a possible state
+            // update and then calls the onSubmit callback.
+            if (this.props.shouldSubmitOnComplete && finalInput.length === CONST.MAGIC_CODE_LENGTH) {
+                setTimeout(() => this.props.onSubmit(finalInput), 0);
```

Comment:
Basically the state updates after the onSubmit logic runs and it uses the previous value (with 5 digits instead of 6). The setTimeout basically waits for the next loop, It's not a clean solution.
Another way would be to use the value passed to the submit on the `validateAndSubmitForm`, but that is called in other places as well, or add a function in between that calls the `validateAndSubmitForm` with the passed value (this last one is my favorite). What do you think?

Reply:
You mean inside the setState of the magicCodeInput? The problem will remain, the problem is not with the state inside the MagicCodeInput but with the `validateCode` state in `BaseValidateCodeForm`, which has not finished updated when the onSubmit -> validateAndSubmitForm is called.